### PR TITLE
 implement VFixnumField

### PR DESCRIPTION
**Pull request by SNsir**

### CODE REVIEW

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 1]
```diff
@@ -18,12 +18,809 @@
...
+  fun toText(v: Fixed): String {
+    return if (!fraction) {
+      v.toString()
+    } else {
+      toFraction(v.toString())
+    }
+  }
+
+  private fun toFraction(str: String): String {
+    var dot: Int
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:518]: val dot

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 2]
```diff
@@ -18,12 +18,809 @@
...
+    if ((str.indexOf(',').also { dot = it }) == -1) {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:519]: Add empty line before if

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 3]
```diff
@@ -18,12 +18,809 @@
...
+      return str
+    }
+    val precomma = str.substring(0, dot)
+    val fract = Integer.valueOf(str.substring(dot + 1, str.length)).toInt()
+    when {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:524]: empty line between

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 4]
```diff
@@ -18,12 +18,809 @@
...
+      fract * 64 % 1000000 != 0 -> {
+        return str
+      }
+      fract == 0 -> {
+        return precomma
+      }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:530]: Remove {} : 
... -> return ..

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 5]
```diff
@@ -18,12 +18,809 @@
...
+          precomma == "0" -> {
+            return "$num/$den"
+          }
+          precomma == "-0" -> {
+            return "-$num/$den"
+          }
+          else -> {
+            return "$precomma $num/$den"
+          }
+        }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:549]: Refactor to : 
return when (precomma) {
          "0" -> "$num/$den"
          "-0" -> "-$num/$den"
          else -> "$precomma $num/$den"
        }

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 6]
```diff
@@ -18,12 +18,809 @@
...
+      if (value[0] != null && value[0]!! < criticalMinValue!!) {
+        setHasCriticalValue(true)
+        return
+      }
+    }
+    if (criticalMaxValue != null) {
+      if (value[0] != null && value[0]!!.compareTo(criticalMaxValue!!) > 0) {
+        setHasCriticalValue(true)
+        return
+      }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:565]: Refactor to : 
if (value[0] != null 
          && (criticalMinValue != null 
          && value[0]!! < criticalMinValue!! || criticalMaxValue != null 
                                                && value[0]!! > criticalMaxValue!!)) {
        setHasCriticalValue(true)
        return
      }

File: galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt [hunk 7]
```diff
@@ -15,6 +15,273 @@
...
+
+  open fun getSqlImpl(r: Int): String? {
+    TODO()
+  }
+
+  open fun toObject(s: String): Any? {
+    TODO()
+  }
+
+  open fun getDisplay(): UField {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VField.kt:172]: Return type is "UField?"

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 8]
```diff
@@ -18,12 +18,809 @@
...
+    }
+    setHasCriticalValue(false)
+  }
+
+  private fun setHasCriticalValue(critical: Boolean) {
+    if (getDisplay() != null) {
+      (getDisplay() as UTextField).setHasCriticalValue(critical)
+    }
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:573]: Change to : 
getDisplay()?.let { (it as UTextField).setHasCriticalValue(critical) }

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 9]
```diff
@@ -18,12 +18,809 @@
...
+  }
+
+  /**
+   * Returns `true` if its is a fraction field.
+   * @return `true` if its is a fraction field.
+   */
+  /*
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:580]: Delete comment

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 10]
```diff
@@ -18,12 +18,809 @@
...
+     * ----------------------------------------------------------------------
+     * DATA MEMBERS
+     * ----------------------------------------------------------------------
+     */
+
+  // static (compiled) data
+  val fraction // display as fraction
+          : Boolean
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:588]: Fix comment position

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 11]
```diff
@@ -18,12 +18,809 @@
...
+  private val digits: Int
+
+  /**
+   * Represents the max permitted value.
+   * @return The max permitted value.
+   */
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:594]: Delete comment

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 12]
```diff
@@ -18,12 +18,809 @@
...
+import org.kopi.galite.visual.MessageCode
+import org.kopi.galite.visual.VlibProperties
 
-//TODO
-class VFixnumField {
+class VFixnumField(digits: Int,
+                   maxScale: Int,
+                   fraction: Boolean,
+                   var minval: Fixed?,
+                   var maxval: Fixed?)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:36]: private val digits: Int,
private var maxScale: Int,
val fraction: Boolean,
var minval: Fixed?,
var maxval: Fixed?

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 13]
```diff
@@ -33,7 +830,49 @@ class VFixnumField {
...
+  /**
+   * Constructor
+   */
+  init {
+    fieldMaxScale = maxScale
+    this.maxScale = maxScale
+    minval = minval?.setScale(fieldMaxScale) ?: calculateUpperBound(digits, fieldMaxScale).negate()
+    maxval = maxval?.setScale(fieldMaxScale) ?: calculateUpperBound(digits, fieldMaxScale)
+    this.digits = digits
+    this.fraction = fraction
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:874]: We dont need to add this.property = arg when the property is already declared as val or var in the constructor

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 14]
```diff
@@ -33,7 +830,49 @@ class VFixnumField {
...
+      }
+    }
+  }
+
+  /**
+   * Constructor
+   */
+  init {
+    fieldMaxScale = maxScale
+    this.maxScale = maxScale
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:870]: Delete

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 15]
```diff
@@ -18,12 +18,809 @@
...
+
+  protected var criticalMinValue: Fixed? = null
+
+  protected var criticalMaxValue: Fixed? = null
 
   companion object {
+    /*
+   * ----------------------------------------------------------------------
+   * PRIVATE METHODS
+   * ----------------------------------------------------------------------
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:618]: Delete

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 16]
```diff
@@ -18,12 +18,809 @@
...
+      } else {
+        if (negative) {
+          value = -value
+        }
+        NotNullFixed(value, scale)
+      }
+    }
+
+    /**
+     * Calculaes the upper bound of a fixnum field : FIXNUM(digits, scale)
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:802]: Calculates

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VlibProperties.kt [hunk 17]
```diff
@@ -19,6 +19,10 @@
 package org.kopi.galite.visual
 
 object VlibProperties {
+  fun getString(key: String, arrayOf: Array<Any>): String {
```
Comment (ZMarouani) [inline galite-core/src/main/kotlin/org/kopi/galite/visual/VlibProperties.kt:22]: I think it needs to be Array<Any?>

**Review by ZMarouani — CHANGES_REQUESTED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/visual/VlibProperties.kt (hunk 17 above)
Comment (SNsir) [reply galite-core/src/main/kotlin/org/kopi/galite/visual/VlibProperties.kt:22]: ignore it

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 18]
```diff
@@ -18,12 +18,809 @@
...
+
+  /**
+   * Sets the field value of given record.
+   *
+   * Warning:   This method will become inaccessible to kopi users in next release
+   * @kopi      inaccessible
+   */
+  override fun setObject(r: Int, v: Any) {
+    // !!! HACK for Oracle
+    if (v != null && (v is Int)) {
```
[the remark this answers is not in the document]
Comment (SNsir) [reply galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:375]: it is an overriden method, can't be nullable

**Review by ZMarouani — APPROVED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 19]
```diff
@@ -282,10 +284,10 @@ class VFixnumField(private val digits: Int,
    */
   fun setMaxScale(scale: Int) {
     // dynamic maxScale mustn't exceed the maxScale defined in the field declaration (fieldMaxScale).
-    if (scale > fieldMaxScale) {
+    if (scale > maxScale) {
```
Comment (h-haddad) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:287]: maxScale --> fieldMaxScale 

then in  fun setScale(record: Int, scale: Int) { --> change fieldMaxScale to maxScale

**Review by h-haddad — APPROVED** [review]

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 20]
```diff
@@ -18,12 +18,784 @@
...
+    getDisplay().let { (it as UTextField).setHasCriticalValue(critical) }
+  }
+
+  /*
+     * ----------------------------------------------------------------------
+     * DATA MEMBERS
+     * ----------------------------------------------------------------------
+     */
+
+  // dynamic data
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:571]: move this just before `minval`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 21]
```diff
@@ -18,12 +18,784 @@
...
+   * Sets the maxScale value for the current record.
+   *
+   * @param     scale           the scale value.
+   */
+  fun setMaxScale(scale: Int) {
+    // dynamic maxScale mustn't exceed the maxScale defined in the field declaration (fieldMaxScale).
+    if (scale > fieldMaxScale) {
+      throw InconsistencyException(MessageCode.getMessage("VIS-00060",
+                                                           scale,
+                                                           fieldMaxScale))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:292]: fieldMaxScale.toString()

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 22]
```diff
@@ -18,12 +18,784 @@
...
+
+  /**
+   * Sets the scale value for the specified record.
+   *
+   * @param     scale           the scale value.
+   * @param     record          the record value.
+   */
+  fun setScale(record: Int, scale: Int) {
+    if (scale > maxScale) {
+      throw InconsistencyException(MessageCode.getMessage("VIS-00060", scale, maxScale))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:267]: scale.toString(), maxScale.toString()

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 23]
```diff
@@ -18,12 +18,784 @@
...
+      val v: Fixed?
+
+      try {
+        v = scanFixed(s)
+      } catch (e: NumberFormatException) {
+        throw VFieldException(this, MessageCode.getMessage("VIS-00006"))
+      }
+      if (v != null) {
+        if (v.scale > scale) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00011", arrayOf<Any>(scale) as Array<Any>?))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:433]: arrayOf(scale)

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 24]
```diff
@@ -18,12 +18,784 @@
...
+        }
+        if (minval != null && v.compareTo(minval!!) == -1) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00012", arrayOf<Any>(minval!!) as Array<Any>?))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:436]: arrayOf(minVal)

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 25]
```diff
@@ -18,12 +18,784 @@
...
+        }
+        if (maxval != null && v.compareTo(maxval!!) == 1) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00009", arrayOf<Any>(maxval!!) as Array<Any>?))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:439]: arrayOf(maxVal)

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 26]
```diff
@@ -18,12 +18,784 @@
...
-class VFixnumField {
+  /**
+   * Returns the sum of the field values of all records.
+   *
+   * @param     exclude         exclude the current record
+   * @param     coalesceValue   the value to take if all fields are empty
+   * @return    the sum of the field values or coalesceValue if none is filled.
+   */
+  fun computeSum(exclude: Boolean, coalesceValue: NotNullFixed): NotNullFixed
+          = computeSum(exclude)?.let { computeSum(exclude) as NotNullFixed } ?: coalesceValue
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:227]: code style : move '=' to the previous line
`as?`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 27]
```diff
@@ -18,12 +18,787 @@
...
+      val v: Fixed?
+
+      try {
+        v = scanFixed(s)
+      } catch (e: NumberFormatException) {
+        throw VFieldException(this, MessageCode.getMessage("VIS-00006"))
+      }
+      if (v != null) {
+        if (v.scale > scale) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00011", arrayOf(scale) as Array<Any>?))
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:409]: remove ` as Array<Any>?`

File: galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt [hunk 28]
```diff
@@ -18,12 +18,787 @@
...
+        }
+        if (minval != null && v.compareTo(minval) == -1) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00012", arrayOf(minval) as Array<Any>?))
+        }
+        if (maxval != null && v.compareTo(maxval) == 1) {
+          throw VFieldException(this, MessageCode.getMessage("VIS-00009", arrayOf(maxval) as Array<Any>?))
+        }
+        if (toText(v.setScale(maxScale)).length > width) {
```
Comment (hfazai) [inline galite-core/src/main/kotlin/org/kopi/galite/form/VFixnumField.kt:417]: remove ` as Array<Any>?`

**Pull request merged**